system/lib/wasmfs/virtual.h

```diff
@@ -0,0 +1,123 @@
+// Copyright 2022 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+// Utilities for creating virtual backends that forward operations to underlying
+// "real" files. Virtual backends must wrap all underlying files, even if the
+// wrapper is a no-op, so that no part of the system observes mixed backends. To
+// the backend-independent code, the root directory of the virtual backend and
+// all of its descendants must appear to have the same virtual backend. To the
+// underlying "real" backend, all of its files must appear to have parents of
+// the same real backend. Without these invariants, renames would not work
+// correctly and without wrapping all files, these invariants would not be
+// upheld.
+
+#pragma once
+
+#include "file.h"
+
+namespace wasmfs {
+
+static inline std::shared_ptr<File> devirtualize(std::shared_ptr<File>);
+
+class VirtualDataFile : public DataFile {
+protected:
+  std::shared_ptr<DataFile> real;
+
+public:
+  VirtualDataFile(std::shared_ptr<DataFile> real, backend_t backend)
+    : DataFile(real->locked().getMode(), backend), real(real) {}
+
+protected:
+  virtual off_t getSize() override { return real->locked().getSize(); }
+  virtual int open(oflags_t flags) override {
+    return real->locked().open(flags);
+  }
+  virtual int close() override { return real->locked().close(); }
+  virtual ssize_t read(uint8_t* buf, size_t len, off_t offset) override {
+    return real->locked().read(buf, len, offset);
+  }
+  virtual ssize_t write(const uint8_t* buf, size_t len, off_t offset) override {
+    return real->locked().write(buf, len, offset);
+  }
+  virtual int setSize(off_t size) override {
+    return real->locked().setSize(size);
+  }
+  virtual int flush() override { return real->locked().flush(); }
+
+  friend std::shared_ptr<File> devirtualize(std::shared_ptr<File>);
+};
+
+// Forward operations through Directory::Handle to get the proper dcache logic.
+class VirtualDirectory : public Directory {
+protected:
+  std::shared_ptr<Directory> real;
+
+public:
+  VirtualDirectory(std::shared_ptr<Directory> real, backend_t backend)
+    : Directory(real->locked().getMode(), backend), real(real) {}
+
+protected:
+  virtual std::shared_ptr<File> getChild(const std::string& name) override {
+    return real->locked().getChild(name);
```

Comment:
In theory if the virtual thing owned the real thing, we could rely on the virtual thing being locked to imply the real thing is locked, and avoid a lock here..? But that seems tricky.

Reply:
Yes, and it might be possible to express that by having a private method to get a handle without acquiring the lock then using `friend` to make that method visible to `VirtualDataFile` and friends. That complexity doesn't seem worth it for now, though.